File: server/jobs/fire_progression.py
"""Fire stage progression engine — background scheduler.

Runs as an asyncio background task that periodically:
1. Scans all active fire grid keys in Redis
2. Removes expired fire events (TTL-based via sorted set scores)
3. Recalculates fire stage for each grid
4. Broadcasts stage changes to subscribed clients via Socket.IO
5. Broadcasts firefighter:spawn event at stage ≥ 4 (visual effect only)

The scheduler uses Redis Sorted Sets where:
- Key:    fire:{gridId}
- Score:  expiration timestamp (now + FIRE_TTL_SEC at creation)
- Member: unique event ID

Active fires = members with score > current time.
"""

from __future__ import annotations

import asyncio
import logging
import time
from datetime import datetime
from typing import TYPE_CHECKING, Any, Protocol

from config import (
    KST,
    STATS_DAILY_FIRES_PREFIX,
    STATS_DAILY_FIRES_TTL_SEC,
    STATS_DAILY_RANKING_PREFIX,
    STATS_DAILY_RANKING_TTL_SEC,
    STATS_TOTAL_FIRES_KEY,
)
from grid import grid_id_to_center
from models.fire import (
    FireStage,
    build_grid_state,
    get_firefighter_remove_count,
    get_stage,
    should_dispatch_firefighter,
)
from models.firefighter import (
    FirefighterNPC,
    create_firefighter_npc,
    should_spawn_firefighter,
)

if TYPE_CHECKING:
    from redis.asyncio import Redis

    from services.admin_region import AdminRegionResolver

logger = logging.getLogger(__name__)

# Redis key prefix for fire grids
FIRE_KEY_PREFIX = "fire:"

# Key tracking all active grid IDs (a Redis Set)
ACTIVE_GRIDS_KEY = "active_grids"


class Broadcaster(Protocol):
    """Protocol for broadcasting events to Socket.IO rooms."""

    async def broadcast_to_room(self, event: str, data: Any, room: str) -> None: ...
    async def broadcast(self, event: str, data: Any) -> None: ...


class FireProgressionEngine:
    """Time-based auto-escalation scheduler for fire stage progression.

    Manages the lifecycle of fires across all grid cells:
    - Cleans expired fire events from Redis sorted sets
    - Tracks previous stage per grid to detect transitions
    - Broadcasts fire:update on stage change
    - Broadcasts firefighter:spawn event at stage ≥ 4 (visual effect only)

    Args:
        redis: Async Redis client instance.
        broadcaster: Object implementing broadcast_to_room and broadcast.
        scan_interval: Seconds between each progression scan (default: 2s).
        cleanup_interval: Seconds between deep cleanup runs (default: 60s).
    """

    def __init__(
        self,
        redis: Redis,
        broadcaster: Broadcaster,
        scan_interval: float = 2.0,
        cleanup_interval: float = 60.0,
    ) -> None:
        self._redis = redis
        self._broadcaster = broadcaster
        self._scan_interval = scan_interval
        self._cleanup_interval = cleanup_interval

        # Track previous stage per grid to detect transitions
        self._grid_stages: dict[str, FireStage] = {}

        # Admin region resolver for daily ranking (injected at startup; optional)
        self._resolver: "AdminRegionResolver | None" = None

        # Background task references
        self._progression_task: asyncio.Task | None = None
        self._cleanup_task: asyncio.Task | None = None
        self._running = False

    @property
    def is_running(self) -> bool:
        return self._running

    @property
    def grid_stages(self) -> dict[str, FireStage]:
        """Current tracked stage per grid (read-only copy)."""
        return dict(self._grid_stages)

    def set_region_resolver(self, resolver: "AdminRegionResolver | None") -> None:
        """Inject the admin region resolver used for daily ranking.

        Accepts None to disable ranking writes without failing the pipeline.
        """
        self._resolver = resolver

    def start(self) -> None:
        """Start all background loops as asyncio tasks."""
        if self._running:
            logger.warning("FireProgressionEngine already running")
            return

        self._running = True
        self._progression_task = asyncio.create_task(
            self._progression_loop(), name="fire-progression"
        )
        self._cleanup_task = asyncio.create_task(
            self._cleanup_loop(), name="fire-cleanup"
        )
        logger.info(
            "FireProgressionEngine started (scan=%.1fs, cleanup=%.1fs)",
            self._scan_interval,
            self._cleanup_interval,
        )

    async def stop(self) -> None:
        """Gracefully stop all background loops."""
        self._running = False
        tasks = [
            t
            for t in (self._progression_task, self._cleanup_task)
            if t is not None
        ]
        for task in tasks:
            task.cancel()
        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)

        self._progression_task = None
        self._cleanup_task = None
        logger.info("FireProgressionEngine stopped")

    # ------------------------------------------------------------------
    # Core: Stage progression scan
    # ------------------------------------------------------------------

    async def _progression_loop(self) -> None:
        """Main loop: scan grids, detect stage changes, broadcast updates."""
        while self._running:
            try:
                await self._scan_and_update_stages()
            except asyncio.CancelledError:
                break
            except Exception:
                logger.exception("Error in progression loop")
            await asyncio.sleep(self._scan_interval)

    async def _scan_and_update_stages(self) -> None:
        """Scan all active grids and broadcast any stage transitions."""
        now = time.time()
        grid_ids = await self._get_active_grid_ids()

        for grid_id in grid_ids:
            active_count = await self._get_active_count(grid_id, now)
            new_stage = get_stage(active_count)
            prev_stage = self._grid_stages.get(grid_id, FireStage.NONE)

            if new_stage != prev_stage:
                self._grid_stages[grid_id] = new_stage
                await self._broadcast_stage_change(
                    grid_id, active_count, new_stage, prev_stage=prev_stage,
                )

                # Trigger firefighter NPC spawn on escalation to stage 4+
                if new_stage >= FireStage.DAEHWAJAE:
                    await self.check_and_spawn_firefighter(grid_id, active_count)

            # Clean up tracking for empty grids (even if stage didn't change)
            if new_stage == FireStage.NONE:
                self._grid_stages.pop(grid_id, None)
                await self._remove_active_grid(grid_id)

    async def _get_active_count(self, grid_id: str, now: float) -> int:
        """Count active (non-expired) fires in a grid cell.

        Args:
            grid_id: The grid cell identifier.
            now: Current timestamp for expiry comparison.

        Returns:
            Number of active fire events.
        """
        key = f"{FIRE_KEY_PREFIX}{grid_id}"
        # ZCOUNT with score range [now, +inf] = active fires
        count = await self._redis.zcount(key, now, "+inf")
        return count

    async def _get_active_grid_ids(self) -> list[str]:
        """Get all grid IDs that have (or recently had) active fires.

        Uses the active_grids Redis Set to track which grids have fires,
        avoiding expensive SCAN operations.
        """
        members = await self._redis.smembers(ACTIVE_GRIDS_KEY)
        return [m.decode() if isinstance(m, bytes) else m for m in members]

    async def _remove_active_grid(self, grid_id: str) -> None:
        """Remove a grid from the active tracking set."""
        await self._redis.srem(ACTIVE_GRIDS_KEY, grid_id)

    async def _broadcast_stage_change(
        self, grid_id: str, active_count: int, new_stage: FireStage,
        prev_stage: FireStage | None = None,
    ) -> None:
        """Broadcast fire:update and fire:stage_transition events on stage change.

        Sends updates to both the grid room (for viewport subscribers) and
        globally (for map overview clients). The fire:stage_transition event
        carries prev/new stage info so clients can animate transitions.

        Args:
            grid_id: Grid cell that changed.
            active_count: Current active fire count.
            new_stage: The new fire stage.
            prev_stage: The previous fire stage (for transition payload).
        """
        import time as _time

        state = build_grid_state(grid_id=grid_id, active_count=active_count)
        payload = state.model_dump()
        payload["timestamp"] = _time.time()
        # camelCase aliases for mock server compat
        payload["gridId"] = grid_id
        payload["activeCount"] = active_count

        # 1) Room-scoped update — reaches viewport subscribers
        await self._broadcaster.broadcast_to_room("fire:update", payload, room=grid_id)

        # 2) Global update — reaches map overview clients
        await self._broadcaster.broadcast("fire:update", payload)

        # 3) Dedicated stage transition event with prev/new for animations
        if prev_stage is None:
            prev_stage = FireStage.NONE
        transition_payload = {
            "grid_id": grid_id,
            "active_count": active_count,
            "prev_stage": int(prev_stage),
            "new_stage": int(new_stage),
            "stage_info": payload.get("stage_info"),
            "timestamp": payload["timestamp"],
        }
        await self._broadcaster.broadcast_to_room(
            "fire:stage_transition", transition_payload, room=grid_id,
        )
        await self._broadcaster.broadcast("fire:stage_transition", transition_payload)

        logger.info(
            "Stage change: grid=%s stage=%s→%s count=%d",
            grid_id,
            prev_stage.name,
            new_stage.name,
            active_count,
        )

    async def _broadcast_fire_count_update(
        self, grid_id: str, active_count: int, current_stage: FireStage,
    ) -> None:
        """Broadcast fire:update with current count (even within same stage).

        Used after firefighter suppression so clients see the count decreasing
        even when the stage hasn't changed yet.
        """
        import time as _time

        state = build_grid_state(grid_id=grid_id, active_count=active_count)
        payload = state.model_dump()
        payload["timestamp"] = _time.time()
        # camelCase aliases for mock server compat
        payload["gridId"] = grid_id
        payload["activeCount"] = active_count

        await self._broadcaster.broadcast_to_room("fire:update", payload, room=grid_id)
        await self._broadcaster.broadcast("fire:update", payload)

    # ------------------------------------------------------------------
    # Firefighter spawn (visual effect only)
    # ------------------------------------------------------------------

    async def check_and_spawn_firefighter(self, grid_id: str, active_count: int) -> FirefighterNPC | None:
        """Broadcast firefighter:spawn event when a grid reaches stage 4+.

        Visual-effect only — no Redis state is persisted and no suppression
        is applied. Callers already gate by stage transition, so dedup is
        handled at the call site.

        Args:
            grid_id: Grid cell to check.
            active_count: Current active fire count.

        Returns:
            The spawned FirefighterNPC (payload holder), or None if no spawn needed.
        """
        if not should_spawn_firefighter(active_count):
            return None

        npc = create_firefighter_npc(grid_id, active_count)
        await self._broadcast_firefighter_spawn(npc)

        logger.info(
            "Firefighter spawn broadcast: npc_id=%s grid=%s stage=%d",
            npc.npc_id,
            grid_id,
            npc.target_stage,
        )

        return npc

    async def _broadcast_firefighter_spawn(self, npc: FirefighterNPC) -> None:
        """Broadcast firefighter:spawn event when a new NPC is dispatched."""
        payload = npc.to_broadcast_payload()
        # Broadcast to the specific grid room and also globally
        await self._broadcaster.broadcast_to_room(
            "firefighter:spawn", payload, room=npc.grid_id
        )
        await self._broadcaster.broadcast("firefighter:spawn", payload)

    # ------------------------------------------------------------------
    # Cleanup: remove expired entries
    # ------------------------------------------------------------------

    async def _cleanup_loop(self) -> None:
        """Periodically remove expired fire events from all grid keys."""
        while self._running:
            try:
                await self._run_cleanup()
            except asyncio.CancelledError:
                break
            except Exception:
                logger.exception("Error in cleanup loop")
            await asyncio.sleep(self._cleanup_interval)

    async def _run_cleanup(self) -> None:
        """Remove expired events and clean up empty grid keys."""
        now = time.time()
        grid_ids = await self._get_active_grid_ids()
        cleaned = 0

        for grid_id in grid_ids:
            key = f"{FIRE_KEY_PREFIX}{grid_id}"

            # Remove members with score < now (expired)
            removed = await self._redis.zremrangebyscore(key, "-inf", now)
            if removed > 0:
                cleaned += removed

            # Check if key is now empty
            remaining = await self._redis.zcard(key)
            if remaining == 0:
                await self._redis.delete(key)
                await self._remove_active_grid(grid_id)
                self._grid_stages.pop(grid_id, None)

        if cleaned > 0:
            logger.info("Cleanup: removed %d expired fire events", cleaned)

    # ------------------------------------------------------------------
    # Public helpers (for use by fire API when adding new fires)
    # ------------------------------------------------------------------

    async def register_fire(self, grid_id: str, event_id: str, expire_at: float) -> int:
        """Register a new fire event and return updated active count.

        Called by the fire API when a user ignites a fire.

        Args:
            grid_id: Grid cell where the fire was ignited.
            event_id: Unique fire event identifier.
            expire_at: Expiration timestamp (score in sorted set).

        Returns:
            Current active fire count after registration.
        """
        key = f"{FIRE_KEY_PREFIX}{grid_id}"
        now = time.time()

        # Add fire event to sorted set
        await self._redis.zadd(key, {event_id: expire_at})

        # Track this grid as active
        await self._redis.sadd(ACTIVE_GRIDS_KEY, grid_id)

        # Increment cumulative fire counter
        await self._redis.incr(STATS_TOTAL_FIRES_KEY)

        # Increment today's fire counter (KST-based key, 48h TTL for safe rollover)
        today_str = datetime.now(KST).strftime("%Y-%m-%d")
        today_key = f"{STATS_DAILY_FIRES_PREFIX}{today_str}"
        await self._redis.incr(today_key)
        await self._redis.expire(today_key, STATS_DAILY_FIRES_TTL_SEC)

        # Increment today's region ranking (skip if resolver absent or point
        # falls outside any admin polygon — cumulative/daily counters stay intact)
        if self._resolver is not None:
            lat, lng = grid_id_to_center(grid_id)
            region = self._resolver.resolve(lat, lng)
            if region:
                rank_key = f"{STATS_DAILY_RANKING_PREFIX}{today_str}"
                await self._redis.zincrby(rank_key, 1, region)
                await self._redis.expire(rank_key, STATS_DAILY_RANKING_TTL_SEC)

        # Get current active count
        active_count = await self._redis.zcount(key, now, "+inf")

        # Update stage tracking
        new_stage = get_stage(active_count)
        prev_stage = self._grid_stages.get(grid_id, FireStage.NONE)
        self._grid_stages[grid_id] = new_stage

        # Broadcast immediately on stage change (don't wait for scan)
        if new_stage != prev_stage:
            await self._broadcast_stage_change(
                grid_id, active_count, new_stage, prev_stage=prev_stage,
            )

            # Spawn firefighter NPC on immediate escalation to stage 4+
            if new_stage >= FireStage.DAEHWAJAE:
                await self.check_and_spawn_firefighter(grid_id, active_count)

        return active_count

File: server/internal/engine/fire_progression.go
// Package engine implements background scheduling tasks for fire lifecycle management.
//
// FireProgressionEngine periodically scans all active fire grids in Redis,
// detects fire stage transitions, and broadcasts updates to connected clients
// via Socket.IO. It mirrors the Python server/jobs/fire_progression.py behavior.
package engine

import (
	"context"
	"fmt"
	"log"
	"math"
	"sync"
	"time"

	"github.com/homepy/hwarr/server/internal/grid"
	"github.com/homepy/hwarr/server/internal/model"
)

// RedisProgressionReader abstracts the Redis operations needed by the progression scan.
type RedisProgressionReader interface {
	// SMembers returns all members of the set at key.
	SMembers(ctx context.Context, key string) ([]string, error)
	// ZCount counts sorted set members with score between min and max (inclusive, string-encoded).
	ZCount(ctx context.Context, key, min, max string) (int64, error)
	// SRem removes members from the set at key.
	SRem(ctx context.Context, key string, members ...interface{}) (int64, error)
}

// Broadcaster abstracts Socket.IO broadcasting operations.
type Broadcaster interface {
	// BroadcastToRoom sends an event to all sockets in a specific room.
	BroadcastToRoom(namespace, room, event string, args ...interface{}) (int, error)
	// BroadcastToNamespace sends an event to all sockets in a namespace.
	BroadcastToNamespace(namespace, event string, args ...interface{}) (int, error)
}

// FireProgressionEngine is a background scheduler that periodically:
//  1. Scans all active fire grid keys in Redis (2-second default interval)
//  2. Recalculates fire stage for each grid based on active (non-expired) fire count
//  3. Detects stage transitions and broadcasts fire:update + fire:stage_transition
//  4. Broadcasts firefighter:spawn event when stage >= 4 (visual effect only)
//
// This is a 1:1 port of Python server/jobs/fire_progression.py's progression loop.
type FireProgressionEngine struct {
	redis       RedisProgressionReader
	broadcaster Broadcaster
	logger      *log.Logger

	scanInterval time.Duration

	// mu protects gridStages
	mu         sync.RWMutex
	gridStages map[string]model.FireStage

	// running state
	running bool
	cancel  context.CancelFunc
	wg      sync.WaitGroup
}

// NewFireProgressionEngine creates a new engine.
//
// Parameters:
//   - redis: Redis client for reading fire data
//   - broadcaster: Socket.IO server for broadcasting events
//   - scanInterval: interval between progression scans (default 2s)
func NewFireProgressionEngine(
	redis RedisProgressionReader,
	broadcaster Broadcaster,
	scanInterval time.Duration,
	logger *log.Logger,
) *FireProgressionEngine {
	if logger == nil {
		logger = log.Default()
	}
	if scanInterval <= 0 {
		scanInterval = 2 * time.Second
	}

	return &FireProgressionEngine{
		redis:        redis,
		broadcaster:  broadcaster,
		logger:       logger,
		scanInterval: scanInterval,
		gridStages:   make(map[string]model.FireStage),
	}
}

// IsRunning returns whether the engine is currently running.
func (e *FireProgressionEngine) IsRunning() bool {
	e.mu.RLock()
	defer e.mu.RUnlock()
	return e.running
}

// GridStages returns a snapshot of the current tracked stage per grid.
func (e *FireProgressionEngine) GridStages() map[string]model.FireStage {
	e.mu.RLock()
	defer e.mu.RUnlock()
	result := make(map[string]model.FireStage, len(e.gridStages))
	for k, v := range e.gridStages {
		result[k] = v
	}
	return result
}

// SetGridStage sets the stage for a grid (used by external fire registration
// to keep the engine's stage tracking in sync without waiting for the next scan).
func (e *FireProgressionEngine) SetGridStage(gridID string, stage model.FireStage) {
	e.mu.Lock()
	defer e.mu.Unlock()
	if stage == model.StageNone {
		delete(e.gridStages, gridID)
	} else {
		e.gridStages[gridID] = stage
	}
}

// GetGridStage returns the current tracked stage for a grid.
func (e *FireProgressionEngine) GetGridStage(gridID string) model.FireStage {
	e.mu.RLock()
	defer e.mu.RUnlock()
	return e.gridStages[gridID]
}

// Start begins the progression scan loop. It is non-blocking and launches
// a goroutine that runs until Stop() is called.
func (e *FireProgressionEngine) Start() {
	e.mu.Lock()
	if e.running {
		e.mu.Unlock()
		e.logger.Println("FireProgressionEngine already running")
		return
	}

	ctx, cancel := context.WithCancel(context.Background())
	e.cancel = cancel
	e.running = true
	e.mu.Unlock()

	e.wg.Add(1)
	go e.progressionLoop(ctx)

	e.logger.Printf("FireProgressionEngine started (scan=%.1fs)", e.scanInterval.Seconds())
}

// Stop gracefully shuts down the progression loop.
func (e *FireProgressionEngine) Stop() {
	e.mu.Lock()
	if !e.running {
		e.mu.Unlock()
		return
	}
	e.running = false
	e.cancel()
	e.mu.Unlock()

	e.wg.Wait()
	e.logger.Println("FireProgressionEngine stopped")
}

// progressionLoop is the main goroutine: scan grids, detect stage changes, broadcast.
func (e *FireProgressionEngine) progressionLoop(ctx context.Context) {
	defer e.wg.Done()

	ticker := time.NewTicker(e.scanInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			if err := e.scanAndUpdateStages(ctx); err != nil {
				if ctx.Err() != nil {
					return // context cancelled, expected
				}
				e.logger.Printf("Error in progression scan: %v", err)
			}
		}
	}
}

// scanAndUpdateStages scans all active grids and broadcasts any stage transitions.
// This is the core logic matching Python's _scan_and_update_stages().
func (e *FireProgressionEngine) scanAndUpdateStages(ctx context.Context) error {
	nowStr := fmt.Sprintf("%d", time.Now().Unix())

	gridIDs, err := e.getActiveGridIDs(ctx)
	if err != nil {
		return fmt.Errorf("get active grids: %w", err)
	}

	for _, gridID := range gridIDs {
		activeCount, err := e.getActiveCount(ctx, gridID, nowStr)
		if err != nil {
			e.logger.Printf("Error counting fires for grid %s: %v", gridID, err)
			continue
		}

		newStage := model.GetStage(activeCount)

		e.mu.RLock()
		prevStage := e.gridStages[gridID]
		e.mu.RUnlock()

		if newStage != prevStage {
			e.mu.Lock()
			e.gridStages[gridID] = newStage
			e.mu.Unlock()

			e.broadcastStageChange(gridID, activeCount, newStage, prevStage)

			// Trigger firefighter NPC spawn on escalation to stage 4+
			if newStage >= model.StageDaehwajae {
				e.broadcastFirefighterSpawn(gridID, activeCount)
			}
		}

		// Clean up tracking for empty grids
		if newStage == model.StageNone {
			e.mu.Lock()
			delete(e.gridStages, gridID)
			e.mu.Unlock()
			_ = e.removeActiveGrid(ctx, gridID)
		}
	}

	return nil
}

// getActiveGridIDs returns all grid IDs in the active_grids Redis set.
func (e *FireProgressionEngine) getActiveGridIDs(ctx context.Context) ([]string, error) {
	return e.redis.SMembers(ctx, ActiveGridsKey)
}

// getActiveCount counts active (non-expired) fires in a grid cell.
// Active fires = sorted set members with score > current time.
func (e *FireProgressionEngine) getActiveCount(ctx context.Context, gridID, nowStr string) (int, error) {
	key := FireKeyPrefix + gridID
	count, err := e.redis.ZCount(ctx, key, nowStr, "+inf")
	if err != nil {
		return 0, err
	}
	return int(count), nil
}

// removeActiveGrid removes a grid from the active tracking set.
func (e *FireProgressionEngine) removeActiveGrid(ctx context.Context, gridID string) error {
	_, err := e.redis.SRem(ctx, ActiveGridsKey, gridID)
	return err
}

// broadcastStageChange broadcasts fire:update and fire:stage_transition events.
// Matches Python _broadcast_stage_change exactly.
func (e *FireProgressionEngine) broadcastStageChange(
	gridID string,
	activeCount int,
	newStage model.FireStage,
	prevStage model.FireStage,
) {
	ts := float64(time.Now().UnixMilli()) / 1000.0

	centerLat, centerLng, cerr := grid.GridIDToCenter(gridID)
	if cerr != nil {
		e.logger.Printf("broadcastStageChange GridIDToCenter error for %s: %v", gridID, cerr)
	}

	state := model.BuildGridState(gridID, activeCount, centerLat, centerLng)
	payload := map[string]interface{}{
		"gridId":      state.GridID,
		"lat":         state.Lat,
		"lng":         state.Lng,
		"activeCount": state.ActiveCount,
		"stage":       state.Stage,
		"stageInfo":   state.StageInfo,
		"timestamp":   ts,
	}

	// Room-scoped update — reaches viewport subscribers only.
	if _, err := e.broadcaster.BroadcastToRoom("/", gridID, "fire:update", payload); err != nil {
		e.logger.Printf("Failed to broadcast fire:update to room %s: %v", gridID, err)
	}

	// Dedicated stage transition event with prev/new for animations (room-scoped).
	transitionPayload := map[string]interface{}{
		"gridId":      gridID,
		"lat":         state.Lat,
		"lng":         state.Lng,
		"activeCount": activeCount,
		"prevStage":   int(prevStage),
		"newStage":    int(newStage),
		"stageInfo":   state.StageInfo,
		"timestamp":   ts,
	}
	if _, err := e.broadcaster.BroadcastToRoom("/", gridID, "fire:stage_transition", transitionPayload); err != nil {
		e.logger.Printf("Failed to broadcast fire:stage_transition to room %s: %v", gridID, err)
	}

	prevName := stageName(prevStage)
	newName := stageName(newStage)
	e.logger.Printf("Stage change: grid=%s stage=%s->%s count=%d", gridID, prevName, newName, activeCount)
}

// broadcastFirefighterSpawn broadcasts firefighter:spawn event (visual effect only).
// Matches Python check_and_spawn_firefighter + _broadcast_firefighter_spawn.
func (e *FireProgressionEngine) broadcastFirefighterSpawn(gridID string, activeCount int) {
	stage := model.GetStage(activeCount)
	cfg := model.StageConfigs[stage]
	if !cfg.TriggersFirefighter {
		return
	}

	npcID := fmt.Sprintf("ff-%s", randomHex(8))
	removePerSweep := cfg.FirefighterRemoveCount
	if removePerSweep < 1 {
		removePerSweep = 1
	}

	centerLat, centerLng, _ := grid.GridIDToCenter(gridID)

	payload := map[string]interface{}{
		"npcId":          npcID,
		"gridId":         gridID,
		"lat":            centerLat,
		"lng":            centerLng,
		"status":         "dispatched",
		"dispatchedAt":   float64(time.Now().UnixMilli()) / 1000.0,
		"firesRemoved":   0,
		"removePerSweep": removePerSweep,
		"targetStage":    int(stage),
	}

	if _, err := e.broadcaster.BroadcastToRoom("/", gridID, "firefighter:spawn", payload); err != nil {
		e.logger.Printf("Failed to broadcast firefighter:spawn to room %s: %v", gridID, err)
	}

	e.logger.Printf("Firefighter spawn broadcast: npcId=%s grid=%s stage=%d", npcID, gridID, int(stage))
}

// BroadcastStageChange is the public version of broadcastStageChange, used by
// external callers (e.g., fire registration) to broadcast immediate stage changes
// without waiting for the next scan.
func (e *FireProgressionEngine) BroadcastStageChange(
	gridID string,
	activeCount int,
	newStage model.FireStage,
	prevStage model.FireStage,
) {
	e.broadcastStageChange(gridID, activeCount, newStage, prevStage)

	if newStage >= model.StageDaehwajae {
		e.broadcastFirefighterSpawn(gridID, activeCount)
	}
}

// stageName returns a human-readable label for a fire stage.
func stageName(stage model.FireStage) string {
	if cfg, ok := model.StageConfigs[stage]; ok {
		return cfg.LabelEn
	}
	return "unknown"
}

// randomHex generates a random hex string of the given length.
// Uses a simple time-based approach (good enough for NPC IDs).
func randomHex(n int) string {
	const hex = "0123456789abcdef"
	// Use UnixNano for entropy
	seed := time.Now().UnixNano()
	b := make([]byte, n)
	for i := range b {
		seed = seed*6364136223846793005 + 1442695040888963407 // LCG
		idx := int(math.Abs(float64(seed))) % len(hex)
		b[i] = hex[idx]
	}
	return string(b)
}
